chainer/sequential.py

```diff
@@ -0,0 +1,400 @@
+                x = layer(x)
+        return x
+
+    def __reduce__(self):
+        n_lambda = 0
+        for layer in self._layers:
+            if callable(layer) and hasattr(layer, '__name__') \
+                    and layer.__name__ == '<lambda>':
+                n_lambda += 1
+
+        if n_lambda > 0:
+            raise ValueError(
+                'This Sequential object has at least one lambda function as '
+                'its component. Lambda function can\'t be pickled, so please '
+                'consider to use functools.partial instead of the lambda '
+                'function or use "dill" which is an external package that '
+                'enables pickling an object including lambda functions intead '
+                'of built-in pickle.')
+        return super(Sequential, self).__reduce__()
+
+    def __str__(self):
+        ret = ''
+        for i, layer in enumerate(self):
+            if isinstance(layer, Sequential):
+                name = layer.__class__.__name__
+                name += '\twhich has {} layers'.format(len(layer))
+            elif isinstance(layer, link.Chain):
+                name = layer.__class__.__name__
+                name += '\tThe structure behind a Chain is determined at '
+                name += 'runtime.'
+            elif isinstance(layer, link.ChainList):
+                name = layer.__class__.__name__
+                name += '\tThe structure behind a ChainList is determined at '
+                name += 'runtime.'
+            elif isinstance(layer, link.Link):
+                name = layer.__class__.__name__
+                param_info = '\t'
+                for param in sorted(layer.params(), key=lambda p: p.name):
+                    param_info += param.name
+                    if param._data[0] is not None:
+                        param_info += str(param._data[0].shape)
+                    else:
+                        param_info += '(None)'
+                    param_info += '\t'
+                name = name + param_info
+            elif isinstance(layer, function.Function):
+                name = layer.__class__.__name__
+            elif isinstance(layer, functools.partial):
+                name = repr(layer)
+            elif layer.__name__ == '<lambda>':
+                name = inspect.getsource(layer).strip()
+            else:
+                name = layer.__name__
+            ret += '{}\t{}\n'.format(i, name)
+        return ret
+
+    def append(self, layer):
+        self.insert(len(self), layer)
+
+    def extend(self, sequential):
+        for layer in sequential:
+            self.append(layer)
+
+    def insert(self, i, layer):
+        if not callable(layer):
+            raise ValueError(
+                'All elements of the argment should be callable. But '
+                'given {} is not callable.'.format(layer))
+
+        self._layers.insert(i, layer)
+        if isinstance(layer, link.Link):
+            self.add_link(layer)
```

Comment:
It looks the order of appearances of links in sequential may not correspond to the order of links registered to ChainList. Is it ok? Consider the following code.
```py
seq = Sequential(link1, link2, link3)
seq.insert(0, link0)
```
In this case, the child links are registered like `seq/0 = link1, seq/1 = link2, seq/2 = link3, seq/3 = link0` while the Sequential looks like `[link0, link1, link2, link3]`.

Reply:
Hmm, right. Is that so problematic? I don't come up with a concrete situation which is problematic caused by this behavior. So if you can allow this, I would say it's OK.

Reply:
I'm concerning serialization. The current spec means two `Sequential` objects may be serialized to different binaries even if they hold the same set of elements with the same parameters and the same order but with different orders on constructing them.

Reply:
For example, `Sequential(link0, link1, link2, link3)` behaves almost same as the above example, but is serialized into incompatible binary.

Reply:
Ah, got it. I'll fix it by changing all the indexes when any insertion occurred.

Reply:
Hm, just using `insert` instead of `add_link` resolves this problem?

Reply:
Ah, no. Change the order of links in `_children` list first, and then I need to change all their names.

Reply:
@beam2d Fixed this behavior. The code below seems OK.

```
In [1]: import chainer
In [2]: import chainer.links as L
In [3]: import chainer.functions as F
In [4]: s = chainer.Sequential(L.Linear(1, 2), F.relu, L.Linear(2, 3), F.relu, L.Linear(3, 4))
In [5]: print(s)
0       Linear  W(2, 1) b(2,)
1       relu
2       Linear  W(3, 2) b(3,)
3       relu
4       Linear  W(4, 3) b(4,)
In [6]: [c.name for c in s.children()]
Out[6]: ['0', '1', '2']
In [7]: s.insert(0, L.Linear(1, 1))
In [8]: [c.name for c in s.children()]
Out[8]: ['0', '1', '2', '3']
In [9]: print(s)
0       Linear  W(1, 1) b(1,)
1       Linear  W(2, 1) b(2,)
2       relu
3       Linear  W(3, 2) b(3,)
4       relu
5       Linear  W(4, 3) b(4,)
```